### PR TITLE
Reintroduce FNV hashing

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes: #7131

### Context
Contains changes from:
* https://github.com/dotnet/msbuild/pull/9721
* Last 2 commits is playing it save by putting the calling behind the changewave + preventing inlining - so that when opted out (from 17_10 ChangeWave), the new method body would not be loaded - and hence new StringTools functions would not be JITted. This way opting out of changewave can save anyone who is combining old version of StringTools with new version of Microsoft.Build

### Usage

```
<!-- Works unchanged. Identical to [MSBuild]::StableStringHash($x, 'Legacy') -->
[MSBuild]::StableStringHash($x)

<!-- 
  $hashAlgo will currently allow:
    'Legacy' - the legacy behavior (mimicking string.GetHashCode)
    'Fnv1a32bit' - Fawler-Noll-Vo 1a 32bit
    'Fnv1a32bitFast' - Custom, faster, Fawler-Noll-Vo 1a 32bit
    'Fnv1a64bit' - Fawler-Noll-Vo 1a 64bit
    'Fnv1a64bitFast' -  Custom, faster, Fawler-Noll-Vo 1a 64bit
    'Sha256' - hex string of the Sha256 hash of the given string
-->
[MSBuild]::StableStringHash($x, $hashAlgo)
```

### Testing
* Existing test on colissions extended for all overloads
* Added test on expected output types

### Documentation

https://learn.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2022#msbuild-stablestringhash

### CODE REVIEW

File: documentation/wiki/ChangeWaves.md [hunk 1]
```diff
@@ -83,3 +83,4 @@ A wave of features is set to "rotate out" (i.e. become standard functionality) t
 - [Add Microsoft.IO.Redist for directory enumeration](https://github.com/dotnet/msbuild/pull/6771)
 - [Process-wide caching of ToolsetConfigurationSection](https://github.com/dotnet/msbuild/pull/6832)
 - [Normalize RAR output paths](https://github.com/dotnet/msbuild/pull/6533)
+- [Introduce [MSBuild]::StableStringHash overloads](https://github.com/dotnet/msbuild/issues/9519)
```
Comment (ladipro) [inline documentation/wiki/ChangeWaves.md:86]: Looks like the wrong section of the file. Should be under 17.**10**.

File: src/Build/Evaluation/IntrinsicFunctions.cs [hunk 2]
```diff
@@ -397,12 +399,60 @@ internal static string ConvertFromBase64(string toDecode)
...
-        internal static int StableStringHash(string toHash)
-        {
-            return CommunicationsUtilities.GetHashCode(toHash);
+        internal static object StableStringHashLegacy(string toHash)
+            => CommunicationsUtilities.GetHashCode(toHash);
+
+        /// <summary>
+        /// Hash the string independent of bitness, target framework and default codepage of the environment.
+        /// We do not want this to be inlined, as then the Expander would call directly the new overload, and hence
+        ///  JIT load the functions from StringTools - so we would not be able to prevent their loading with ChangeWave as we do now.
```
Comment (ladipro) [inline src/Build/Evaluation/IntrinsicFunctions.cs:429]: ❤️

Comment (JanKrivanek) [reply src/Build/Evaluation/IntrinsicFunctions.cs:429]: credit goes to @rokonec

File: documentation/wiki/ChangeWaves.md (hunk 1 above)
Comment (JanKrivanek) [reply documentation/wiki/ChangeWaves.md]: Oh my... Thanks forthe catch!

**Review by ladipro — APPROVED** [review]

**Review by rokonec — APPROVED** [review]

**Pull request merged**